src/chart/model/scale.model.ts

```diff
@@ -264,24 +263,41 @@ export class ScaleModel extends ViewportModel {
 		if (initialState.yStart === yStart && initialState.yEnd === yEnd && initialState.zoomY > 0) {
 			return;
 		}
-		super.setYScale(yStart, yEnd, fire);
-		const state = this.export();
-		const constrainedState = this.scalePostProcessor(initialState, state);
 
 		if (this.state.lockPriceToBarRatio) {
-			changeXToKeepRatio(constrainedState, this.zoomXYRatio);
+			this.setLockPriceScale(yStart, yEnd, fire, initialState);
+			return;
+		}
+		this.setYScaleCommon(yStart, yEnd, fire, initialState);
+	}
 
-			this.setXScale(constrainedState.xStart, constrainedState.xEnd);
-			// TODO: rewrite logic for applying constraints to consider both axes, now constraints on Y may not work correctly
+	private setLockPriceScale(yStart: Unit, yEnd: Unit, fire = false, initialState: ViewportModelState) {
+		const zoomIn = yEnd < initialState.yEnd;
+		if ((this.zoomReached.min && zoomIn === false) || (this.zoomReached.max && zoomIn === true)) {
 			return;
-		} else {
-			if (this.state.auto) {
-				this.autoScaleModel.doAutoYScale(constrainedState);
-			}
+		}
 
-			this.apply(constrainedState);
+		super.setYScale(yStart, yEnd, fire);
+		const state = this.export();
+		const constrainedState = this.scalePostProcessor(initialState, state);
+
+		changeXToKeepRatio(initialState, constrainedState);
+		this.zoomReached = this.calculateZoomReached(constrainedState.zoomX, zoomIn);
+		this.setXScale(constrainedState.xStart, constrainedState.xEnd, true, false);
+		this.fireChanged();
+	}
+
+	private setYScaleCommon(yStart: Unit, yEnd: Unit, fire = false, initialState: ViewportModelState) {
```

Comment:
actually I don't see a reason to move this logic in a separate method. It used only once and I don't think that it helps  understanding what's going on, it even may be more misleading for us